File: Day 26 DSA (34)/Code.py
class Solution(object):
    def searchRange(self, nums, target):
        left = self.find(nums, target, True)
        if left == -1:
            return [-1, -1]
        right = self.find(nums, target, False)
        return [left, right]
    
    def find(self, nums, target, isLeft):
        left, right = 0, len(nums) - 1
        a = -1
        
        while left <= right:
            mid = left + (right - left) // 2
            if nums[mid] == target:
                a = mid
                if isLeft:
                    right = mid - 1
                else:
                    left = mid + 1
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
                
        return a

File: Day 26 DSA (34)/Code.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int left = find(nums, target, true);
        if (left == -1) {
            return {-1, -1}; 
        }
        int right = find(nums, target, false);
        return {left, right};
    }
    
private:
    int find(const vector<int>& nums, int target, bool isLeft) {
        int left = 0, right = nums.size() - 1;
        int a = -1;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) {
                a = mid;
                if (isLeft) {
                    right = mid - 1;  // for the left side
                } else {
                    left = mid + 1;   // for right side
                }
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        
        return a;
    }
};
